### ISSUE TITLE
[HOLD for payment 2023-05-16] [$1000]  Details sidebar auto-closes whenever the user clicks on the “Decline” button in one of the multiple IOUs

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#16508]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open an existing chat/new chat
2. Click on “Actions” button
3. Click on “Request Money” link
4. Repeat step 1 to step 3 to create multiple money requests
5. Sign in with the request recipient’s account
6. Open a chat where the user received Money Request messages 
7. Click on “Pay” button
8. Click on “Decline” button
9. Observe that Detials sidebar auto-closes on clicking the “Decline” button 
10. Repeat step 7 to step 8 until you finish declining all money requests

## Expected Result:
Details sidebar should only auto-close once the user finishes declining _all_ the money requests or settling _all_ the requests

## Actual Result:
Details sidebar auto-closes every time the user clicks on the “Decline” button forcing the user to reopen the sidebar to process another request

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.89-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/227728800-c228e3f3-3658-49da-863d-71182882862e.mp4

https://user-images.githubusercontent.com/43996225/227728809-72210ac1-0d49-4bd8-902d-3f05ee9c2a9c.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @Natnael-Guchima
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1679729409034739

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01f0767bdc96a60d89</li>
        <li>Upwork Job ID: 1640412749911355392</li>
        <li>Last Price Increase: 2023-03-27</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @sonialiap (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by MelvinBot** [comment]
Triggered auto assignment to @amyevans (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @rushatgabhane (`External`)

**Comment by Natnael-Guchima** [comment]
This issue seems to exist on Android too. 

https://user-images.githubusercontent.com/57330233/228041853-05463126-47d0-4d24-b89d-a396faf615b8.mp4

**Comment by MelvinBot** [comment]
📣 @Natnael-Guchima! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by Natnael-Guchima** [comment]
Contributor details
Your Expensify account email: natnaelguchima12@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~0135b3a84ec2709c21

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by khatch33** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The IOU details sidebar auto-closes on every decline of a payment request as opposed to only once all requests are declined or settled.
### What is the root cause of that problem?
The cause of the problem is the cancelMoneyRequest function in IOU.js. At the end of the function in line 627 the user is rerouted to the report page. This is occurs on every decline as well as every cancel.  
### What changes do you think we should make in order to solve the problem?
Invoking the Navigation.navigate function only if all requests are settled would solve this problem. I was able to fix the issue locally, but however this would also effect the sidebar when a user cancels a request as opposed to decline. My fix applied forced the sidebar to stay open if there are still unsettled IOU's for both decline and cancel actions which I believe is the desired outcome. 


### What alternative solutions did you explore? (Optional)

**Comment by MelvinBot** [comment]
📣 @khatch33! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by khatch33** [comment]
Contributor details
Your Expensify account email: kylehatch99@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/kyleh13

**Comment by MelvinBot** [comment]
⚠️ Missing/invalid email or upwork profile link. Please make sure you add both your Expensify email and Upwork profile link in the format specified.

**Comment by khatch33** [comment]
Contributor details
Your Expensify account email: kylehatch99@gmail.com
Upwork Profile Link: upwork.com/freelancers/kyleh13

**Comment by rushatgabhane** [comment]
@amyevans I agree with the problem. It can be a chore to cancel / decline multiple IOUs because the RHN keeps closing. 

 I like @hoangzinh's [proposal](https://github.com/Expensify/App/issues/16508#issuecomment-1485943763)!

C+ reviewed 🎀 👀 🎀

**Comment by rushatgabhane** [comment]
I'm so sorry @hoangzinh I completley missed that @khatch33 had posted the same proposal above https://github.com/Expensify/App/issues/16508#issuecomment-1485750043

**Comment by rushatgabhane** [comment]
@amyevans I agree with the problem. It can be a chore to cancel / decline multiple IOUs because the RHN keeps closing.

I like @khatch33's [proposal](https://github.com/Expensify/App/issues/16508#issuecomment-1485750043) to fix this issue!

C+ reviewed 🎀 👀 🎀

**Comment by amyevans** [comment]
Reviewing the proposal, also checking in with @trjExpensify on expected behavior since while I was reviewing it, @Julesssss mentioned this may not be a bug but rather intentional.

**Comment by trjExpensify** [comment]
Yep, I agree this isn't a bug. It should remain open for you to take further actions on the report, like settle the ones you aren't declining for example etc.

**Comment by amyevans** [comment]
I spoke with @trjExpensify again and he agrees that this is a bug. The expected behavior is that the sidebar should only automatically close once all outstanding requests have been cancelled/declined.

@khatch33's [proposal](https://github.com/Expensify/App/issues/16508#issuecomment-1485750043) looks good in theory, and tests well when the requests are in USD.

However in testing with foreign currency, the sidebar does not auto close after declining all outstanding requests:

https://user-images.githubusercontent.com/7277067/229167348-e43f2b33-28e8-40c7-850c-91a6d9765b36.mov

It might make sense to proceed with @khatch33's proposal and address that bug as a separate issue. Thoughts @Julesssss?

**Comment by Julesssss** [comment]
The proposal seems good, but I think we probably should update it to resolve the currency issue too.

**Comment by amyevans** [comment]
Sounds good, @khatch33 can you take a look?

FYI, I'll be OOO next week - feel free to reassign the issue as needed to keep things moving @sonialiap

**Comment by khatch33** [comment]
@amyevans Yes I'm on it!

**Comment by khatch33** [comment]
Just to clarify, is this being updated to include foreign currency requests like @Julesssss mentioned? Because the [IOU total doesn't change locally](https://github.com/Expensify/App/pull/12960) when currencies differ, hasOutstandingIOU therefore doesn't update locally either which is why the sidebar didn't close in @amyevans test. 

As an aside, can anybody touch on the expected behavior when there are opposing payment requests that total 0? For example:
-Account A makes 6 separate 20 dollar requests to B totaling 120 dollars.
-Account B makes a request to A for 100 dollars.
-Account A cancels one request leaving 100 dollars in outstanding requests on each side.

Currently this makes chatReport.hasOutstandingRequest false and disables payment ability. I assume this is intentional but am not sure whether that should close the sidebar automatically.

**Comment by MelvinBot** [comment]
@amyevans, @sonialiap, @rushatgabhane Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by Julesssss** [comment]
> As an aside, can anybody touch on the expected behavior when there are opposing payment requests that total 0? For [example](https://github.com/Expensify/App/issues/16508#issuecomment-1495111071):

In this case there are still outstanding requests. But because there is nothing to pay, no action is possible. I don't think we should close the modal in this case because the user may wish to cancel their IOU request and then settle the balance. 

Might this help with the currency issue? I think only closing when there are no outstanding requests might solve that problem as we don't need to compare currencies.

**Comment by khatch33** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
The IOU details sidebar auto-closes on every decline or cancel of a payment request as opposed to only once all requests are declined or settled.
### What is the root cause of that problem?
The cause of the problem is the cancelMoneyRequest function in IOU.js. At the end of the function in line 627 the user is rerouted to the report page. This is occurs on every decline as well as every cancel. 
### What changes do you think we should make in order to solve the problem?
The Navigation.navigate function should be invoked only if there are no further outstanding requests. The chatReport.hasOutstandingIOU property doesn't work well for this check because it is only checking if the total IOU amount is equal to 0. This leads to false when outstanding requests equal each other, and needs currency conversion for foreign currency requests leading to other problems.

To fix the problem I added a new prop to the ReportTransaction component with a Boolean value representing if it is the only rejectable request remaining.  
<img width="905" alt="Screen Shot 2023-04-05 at 1 28 40 PM" src="https://user-images.githubusercontent.com/95315047/230158965-b5f4a4dc-fbbf-407f-9821-221ec48acdad.png">
I then added a parameter to the cancelMoneyRequest which is used to determine if Navigating back to the chat report is needed, and added it as an argument when calling cancelMoneyRequest.
<img width="376" alt="Screen Shot 2023-04-05 at 1 42 40 PM" src="https://user-images.githubusercontent.com/95315047/230160993-1a96a0fd-689c-44b4-92ef-e10a293e18cd.png"><img width="858" alt="Screen Shot 2023-04-05 at 1 38 54 PM" src="https://user-images.githubusercontent.com/95315047/230160058-2eade9d1-f686-4ab0-a5b0-2d1bde2cabab.png">
<img width="578" alt="Screen Shot 2023-04-05 at 1 40 06 PM" src="https://user-images.githubusercontent.com/95315047/230160616-39fe4d87-9187-41fc-b99d-3ce1a055b24e.png">






### What alternative solutions did you explore? (Optional)
Instead of creating a new onlyRejectable prop, it might be a better solution to change the way chatReport.hasOutstandingIOU is calculated. I'm unsure where else that value is utilized so I was hesitant to change it.

**Comment by khatch33** [comment]
@Julesssss Okay I created a new proposal that accounts for the foreign currency issue, apologies if I should have just updated my first one.

> In this case there are still outstanding requests. But because there is nothing to pay, no action is possible. I don't think we should close the modal in this case because the user may wish to cancel their IOU request and then settle the balance.

I'm not sure if it matters, but right now in this case the hasOutstandingRequests value would be false which might cause an error elsewhere.

**Comment by rushatgabhane** [comment]
@khatch33 thanks for the updated proposal! 

@amyevans let's go with it

**Comment by MelvinBot** [comment]
@amyevans @sonialiap @rushatgabhane this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by MelvinBot** [comment]
@amyevans, @sonialiap, @rushatgabhane Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by MelvinBot** [comment]
@amyevans, @sonialiap, @rushatgabhane Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by sonialiap** [comment]
Amy was OOO last week and also today. Should be back tomorrow 🎉

**Comment by khatch33** [comment]
@rushatgabhane @sonialiap @amyevans Sorry to pester, but I'm now waiting to be hired on Upwork officially correct? Or should I have submitted my pull request.

**Comment by Julesssss** [comment]
Hi @khatch33. Normally you would have been assigned by now, but Amy was OOO (me too). 

I think your proposal is good, but we could probably improve the naming of `onlyRejectable`. Personally I think the prop name could be clearer. But I'll leave that to the PR review.

**Comment by MelvinBot** [comment]
📣 @khatch33 You have been assigned to this job by @Julesssss!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by amyevans** [comment]
Thanks for taking a look at the proposal @Julesssss! I agree it looks good. @khatch33 let us know if have any additional process questions as this is your first issue/PR 😄.

**Comment by MelvinBot** [comment]
@amyevans @Julesssss @sonialiap @rushatgabhane @khatch33 this issue is now 3 weeks old. There is one more week left before this issue <a href='https://stackoverflowteams.com/c/expensify/questions/15233/15234#15234'>breaks WAQ</a> and will need to go internal. What needs to happen to get a PR in review this week? Please create a thread in #expensify-open-source to discuss. Thanks!

**Comment by khatch33** [comment]
@rushatgabhane My bad! I'll submit that today ASAP

**Comment by khatch33** [comment]
So in testing I realized there are some problems that happen when declining/canceling multiple requests rapidly. I thought it was a regression on my end and spent a quite a while debugging, but when multiple `cancelMoneyRequests` are sent at once the server doesn't handle it correctly, and some components in the front end are left with wrong values temporarily. The same thing happens if you go offline, decline 4 requests, and then come back on. Before I go ahead and submit my PR could anybody help me clear up whether this needs to be solved in this fix.

**Comment by amyevans** [comment]
@khatch33 Does [this issue](https://github.com/Expensify/App/issues/12775) sound like what you are seeing? If it doesn't, can you include a recording so we can better understand?

**Comment by khatch33** [comment]
Exactly. @amyevans

**Comment by amyevans** [comment]
Okay, then the answer is No, your PR doesn't need to fix that problem 😅

**Comment by khatch33** [comment]
Okay, I just wanted to double check because it it's pretty easy for it to happen in this case.

**Comment by khatch33** [comment]
@rushatgabhane @amyevans Please let me know if I need to do anything else or if this can get reviewed

**Comment by amyevans** [comment]
@khatch33 Just a heads up for next time, the links in the PR under the `Fixed Issues` heading need to be formatted a particular way, and then our automation will properly assign @rushatgabhane and myself to review the PR. No worries though, I fixed it up manually 😄

**Comment by khatch33** [comment]
Ah my bad, thank you for the help.

**Comment by amyevans** [comment]
DMed @khatch33 in Slack

**Comment by amyevans** [comment]
We're going to open this issue back up to external contributors since it appears the submitted PR has been abandoned. However, if we don't have anyone to take over by Monday, I'll handle the PR internally.

**Comment by jayeshmangwani** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The Details sidebar closes on every decline or cancel of a payment request.

### What is the root cause of that problem?
on cancelMoneyRequest function after calling CancelMoneyRequest we are calling the navigation to report screen

https://github.com/Expensify/App/blob/32fbb2c44aeea0e112f8eb0ca9ea4911e160f499/src/libs/actions/IOU.js#L627

### What changes do you think we should make in order to solve the problem?
we should check on the every cancelMoneyRequest function, if the current IOU cancellable request is the last then we should call navigation

To achieve we have to do following

1. pass a new prop that defines that the last cancellable IOU

https://github.com/Expensify/App/blob/32fbb2c44aeea0e112f8eb0ca9ea4911e160f499/src/pages/iou/IOUTransactions.js#L88-L95

we need to pass a new prop shouldNavigateToRoute that check if the rejectableTransactions length is equal to or less then one then we should navigate to report screen

`shouldNavigateToRoute={
https://user-images.githubusercontent.com/35371050/235065411-0aa36a23-0212-4db1-80df-da4afa4d4af4.mp4
.length <= 1}`

2. Pass `shouldNavigateToRoute` prop to shouldNavigateToRoute function

https://github.com/Expensify/App/blob/32fbb2c44aeea0e112f8eb0ca9ea4911e160f499/src/components/ReportTransaction.js#L47-L54

3. Check if the shouldNavigateToRoute prop is true then navigate

https://github.com/Expensify/App/blob/32fbb2c44aeea0e112f8eb0ca9ea4911e160f499/src/libs/actions/IOU.js#L627

```
if(shouldNavigateToRoute) {
    Navigation.navigate(ROUTES.getReportRoute(chatReportID));
}
```


<details>
  <summary>Result</summary>
  
https://user-images.githubusercontent.com/35371050/235065496-6a4025d5-b7ec-41c8-a05e-11b853f2fb95.mp4

</details>

**Comment by jayeshmangwani** [comment]
[PR](https://github.com/Expensify/App/pull/18190) is ready
@rushatgabhane @Julesssss @amyevans

**Comment by jayeshmangwani** [comment]
@sonialiap Thanks for the offer, but offer does not includes bonus to me

**Comment by jayeshmangwani** [comment]
@sonialiap Thanks, I have accepted

**Comment by Julesssss** [comment]
Hey @rushatgabhane, have you accepted payment? If yes we can close this issue :shipit:

**Comment by sonialiap** [comment]
Offer accepted, Rushat has been paid, closing
